tools/wake-format/emitter.cpp

```diff
@@ -71,6 +90,18 @@ auto Emitter::rhs_fmt() {
   // clang-format on
 }
 
+auto place_comment() {
+  // ws COMMENT -> ws COMMENT nl
+  // ws nl COMMENT -> nl COMMENT nl
+  return fmt().fmt_if(has_comment_next,
+                      fmt()
+                          .fmt_if(TOKEN_WS, fmt().next())
+                          .fmt_if_else(TOKEN_NL, fmt().next().nest(fmt().newline()), fmt().space())
+                          .consume_wsnl()
+                          .token(TOKEN_COMMENT)
+                          .consume_wsnl());
+}
```

Comment:
Yeah sorry, the comment explains it if you've  bashed your head against it a lot.

So we can have a comment 

```
( # comment
something
)
```

or 

```
(
# comment 
something
)
```

and we want to maintain the location of the comment.


So the condition here is 

if TOKEN_WS, TOKEN_COMMENT then emit TOKEN_WS, TOKEN_COMMENT, TOKEN_NL
else if  TOKEN_WS, TOKEN_NL, TOKEN_COMMENT then emit TOKEN_NL TOKEN_COMMENT TOKEN_NL

the reason this is a function is because I've already seen other other case outside this PR where this is needed and there is very likely to be more


The `nest` in there _does_ cause some weirdness that needs to be worked out. In this I did a weird thing in walk_paren but it could be solve by passing a "nest formatter" in or maybe a bool "should nest"

Reply:
I'm a bit confused why the `nest` is needed? Can you explain that?

Reply:
yeah sorry.

Consider 

```
def x = (
    # comment
   1 + 2
)
```

without `nest()`

```
def x = (
# comment
    1 + 2
)
```
would be emitted.

Basically `place_comment()` is assuming that if it inserts a newline it also needs to nest. This turns out to be a bad assumption generally though so I'm thinking it may be better to have a "comment_should_be_newlined" predicate that then leaves the callsite formatter to determine what to do if a newline is required

Reply:
Something isn't right here. This would cause multiple levels of nesting which would result in a parse error. You shouldn't be nesting inside of parens ever. You only nest for things like def/require/match/case/publish.

The problem here isn't with parens its with `def`. The case where the `def` tries not to nest causes this issue and you've smoothed that over using a bunch of hacks. This is frustrating though because we're back to the "failed flat" thing we got rid of.

Reply:
I'm not sure I fully understand.

From my perspective I wasn't smoothing over anything. I just wanted a multiline `paren` node to have its contents nested in by one. Is that not the correct goal?

Assuming is parses, doing 

```
def x = ( 
# comment 
1 + 2
)
```

is also very easy. I just don't like the way to looks

Reply:
If the goal is 

```
def x =
    (
    # comment
    1 + 2
    )
```

then I think I see what your saying. That a different issue inside def though

Reply:
I just wasn't clear on how wake was parsing something until we talked, I think you're outputting things correctly as discussed but I'm not happy with the un-nest stuff. Looking forward to your solution!